### PR TITLE
kv/client: exit region worker if there is no region maintained

### LINKED ISSUE #2284: kv client v2 EventFeed RPC leaks
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do? If possible, provide a recipe for reproducing the error.

Replicating 6k tables using the commit d2c1aaeb81561827cb67ceb6cd51cacc705e702a, we find EventFeed rpc and its goroutine leaks.

From metrics we find, the leakage relates to eventfeed errors.

image

Based on pprof/trace, we find leaked goroutines are spawned at 

https://github.com/pingcap/ticdc/blob/d2c1aaeb81561827cb67ceb6cd51cacc705e702a/cdc/kv/client.go#L855-L860
https://github.com/pingcap/ticdc/blob/d2c1aaeb81561827cb67ceb6cd51cacc705e702a/cdc/kv/region_worker.go#L546-L551

2. What did you expect to see?

No leakage.

3. What did you see instead?

Leakage.

4. Versions of the cluster

A fork of v5.1.0 d2c1aaeb81561827cb67ceb6cd51cacc705e702a

### PR DESCRIPTION
**Pull request by amyangfei**
### What problem does this PR solve? 

Fix #2284

When there is no region maintained in a region worker, we should exit this region worker and release related resource

### What is changed and how it works?

- Add a check that whether there is no region maintained when `handleSingleRegionError`
  - If no region, cancel the gRPC stream, and exit the region worker
- Change the region state manage mechanism, remove region state if region meets error.

Note: clientV1 also has this problem, I will fix on release branch in another PR.

### Check List 

Tests 

 - Unit test
 - Integration test

### Release note

- Reduce goroutine usage when a table's region transfer away from a TiKV node

### CODE REVIEW

**Comment by overvenus** [comment]
Tested with shuffle-leader/region-scheduler, I can confirm it fixes goroutine leakage, but not EventFeed RPC leakage, I can still see EventFeeds increase dramatically.

We need to fix RPC leakage too because TiKV has an upper limit of concurrent running RPCs.

image
image
image

**Comment by amyangfei** [comment]
> Tested with shuffle-leader/region-scheduler, I can confirm it fixes goroutine leakage, but not EventFeed RPC leakage, I can still see EventFeeds increase dramatically.
> 
> We need to fix RPC leakage too because TiKV has an upper limit of concurrent running RPCs.
> 
> image
> image
> image

gRPC stream is closed before region worker exits, PR description includes it

**Comment by amyangfei** [comment]
> > Tested with shuffle-leader/region-scheduler, I can confirm it fixes goroutine leakage, but not EventFeed RPC leakage, I can still see EventFeeds increase dramatically.
> > We need to fix RPC leakage too because TiKV has an upper limit of concurrent running RPCs.
> 
> Yes, gRPC stream is closed before region worker exits, PR description includes it

Note `grpc_client_started_total` in go-grpc-prometheus is increasing only, I didn't find it can decrease. ref: https://github.com/grpc-ecosystem/go-grpc-prometheus/search?q=clientStartedCounter

**Comment by overvenus** [comment]
> > > Tested with shuffle-leader/region-scheduler, I can confirm it fixes goroutine leakage, but not EventFeed RPC leakage, I can still see EventFeeds increase dramatically.
> > > We need to fix RPC leakage too because TiKV has an upper limit of concurrent running RPCs.
> > 
> > 
> > Yes, gRPC stream is closed before region worker exits, PR description includes it
> 
> Note `grpc_client_started_total` in go-grpc-prometheus is increasing only, I didn't find it can decrease. ref: https://github.com/grpc-ecosystem/go-grpc-prometheus/search?q=clientStartedCounter
> Could you find other ways to confirm gRPC connection count @overvenus

Yes, there is a metrics records completed RPCs. `grpc_client_handled_total`

I add metric queries

- rpc: `sum(grpc_client_started_total{tidb_cluster="$tidb_cluster", grpc_method="EventFeed"}) by (instance) - sum(grpc_client_handled_total{tidb_cluster="$tidb_cluster", grpc_method="EventFeed"}) by (instance)`
- rpc-started: `sum(grpc_client_started_total{tidb_cluster="$tidb_cluster", grpc_method="EventFeed"}) by (instance)`
- rpc-handled: `sum(grpc_client_handled_total{tidb_cluster="$tidb_cluster", grpc_method="EventFeed"}) by (instance)`

image
image

The metrics show that it still leak RPC.

**Comment by amyangfei** [comment]
> > > > Tested with shuffle-leader/region-scheduler, I can confirm it fixes goroutine leakage, but not EventFeed RPC leakage, I can still see EventFeeds increase dramatically.
> > > > We need to fix RPC leakage too because TiKV has an upper limit of concurrent running RPCs.
> > > 
> > > 
> > > Yes, gRPC stream is closed before region worker exits, PR description includes it
> > 
> > 
> > Note `grpc_client_started_total` in go-grpc-prometheus is increasing only, I didn't find it can decrease. ref: https://github.com/grpc-ecosystem/go-grpc-prometheus/search?q=clientStartedCounter
> > Could you find other ways to confirm gRPC connection count @overvenus
> 
> Yes, there is a metrics records completed RPCs. `grpc_client_handled_total`
> 
> I add metric queries
> 
> * rpc: `sum(grpc_client_started_total{tidb_cluster="$tidb_cluster", grpc_method="EventFeed"}) by (instance) - sum(grpc_client_handled_total{tidb_cluster="$tidb_cluster", grpc_method="EventFeed"}) by (instance)`
> * rpc-started: `sum(grpc_client_started_total{tidb_cluster="$tidb_cluster", grpc_method="EventFeed"}) by (instance)`
> * rpc-handled: `sum(grpc_client_handled_total{tidb_cluster="$tidb_cluster", grpc_method="EventFeed"}) by (instance)`
> 

> The metrics show that it still leak RPC.

Fixed by release stream resource correctly

**Review by overvenus — APPROVED** [review]
LGTM

File: cdc/kv/client.go [hunk 1]
```diff
@@ -851,6 +851,7 @@ func (s *eventFeedSession) requestRegionToStore(
 
 			limiter := s.client.getRegionLimiter(regionID)
 			g.Go(func() error {
+				defer s.deleteStream(rpcCtx.Addr)
 				if !s.enableKVClientV2 {
 					return s.receiveFromStream(ctx, g, rpcCtx.Addr, getStoreID(rpcCtx), stream, pendingRegions, limiter)
```
Comment (overvenus) [inline cdc/kv/client.go:856]:
```suggestion
					return s.receiveFromStream(streamCtx, g, rpcCtx.Addr, getStoreID(rpcCtx), stream, pendingRegions, limiter)
```

What about streamCtx?

Comment (amyangfei) [reply cdc/kv/client.go:856]: `streamCtx` is used for gRPC stream only, reuse it in `receiveFromStream` could cause unexpected context cancel detection.

**Review by liuzix — APPROVED** [review]

**Pull request merged**